util/clusterauth/clusterauth.go

```diff
@@ -215,39 +216,103 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
 	return GetServiceAccountBearerToken(clientset, ns, ArgoCDManagerServiceAccount)
 }
 
-// GetServiceAccountBearerToken will attempt to get the provided service account until it
-// exists, iterate the secrets associated with it looking for one of type
-// kubernetes.io/service-account-token, and return it's token if found.
+// GetServiceAccountBearerToken determine if a ServiceAccount has a
+// bearer token sercet to use or if one should be created. It then
+// waits for the secret to have a bearer token if a secret needs to
+// be created and returns the token.
 func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
+	secretName, err := createServiceAccountTokenSecret(clientset, sa, ns)
+	if err != nil {
+		return "", err
+	}
+
 	var serviceAccount *corev1.ServiceAccount
 	var secret *corev1.Secret
-	var err error
 	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("Failed to retrieve secret %q: %v", secretName, err)
+		}
+
+		_, ok := secret.Data["token"]
+		if !ok {
+			return false, fmt.Errorf("Secret %q for service account %q did not have a token", secret.Name, serviceAccount)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// createServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func createServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.Poll(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
```

Comment:
Probably needs a timeout.

Reply:
How about 10s? Absolute worst-case scenario, the first call takes 9.999s to return a non-error code, `wait.Poll` calls it once more, and the second call takes 10s. The 10s will let far-away clusters respond, but in 99.9% of cases, low latency will mean we quit trying after ~5s.